src/main/java/pl/cleankod/exchange/core/usecase/AccountServiceImpl.java

```diff
@@ -0,0 +1,85 @@
+
+package pl.cleankod.exchange.core.usecase;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+import pl.cleankod.exchange.core.domain.Account;
+import pl.cleankod.exchange.core.domain.Money;
+import pl.cleankod.exchange.core.gateway.AccountRepository;
+import pl.cleankod.exchange.core.gateway.AccountService;
+import pl.cleankod.exchange.core.gateway.CurrencyConversionService;
+import pl.cleankod.exchange.entrypoint.model.AccountNotFoundException;
+import pl.cleankod.exchange.entrypoint.model.AccountResponse;
+import pl.cleankod.exchange.entrypoint.model.CurrencyConversionException;
+
+import java.util.Currency;
+import java.util.Optional;
+
+@Service
+public class AccountServiceImpl implements AccountService {
+    private final Logger log = LoggerFactory.getLogger(AccountServiceImpl.class);
+    private final AccountRepository accountRepository;
+
+    private final CurrencyConversionService currencyConversionService;
+    private final Currency baseCurrency;
+
+    public AccountServiceImpl(AccountRepository accountRepository, CurrencyConversionService currencyConversionService, Currency baseCurrency) {
+        this.accountRepository = accountRepository;
+        this.currencyConversionService = currencyConversionService;
+        this.baseCurrency = baseCurrency;
+    }
+
+    @Override
+    public AccountResponse findAccountBy(Account.Id id, String targetCurrency, String correlationId) throws AccountNotFoundException {
+
+        return accountRepository.find(id)
+                .map(account -> {
+                    Money money = getMoney(targetCurrency, account, correlationId);
+                    return this.buildAccountDtoFrom(account, money);
+                })
+                .orElseThrow(() -> new AccountNotFoundException(correlationId));
```

Comment:
Do you know another way of handling unhappy flow  scenarios without using stacktraces?